File: AlgoTemplates/FFT.cs
using System;
using System.Collections.Generic;
using System.Numerics;
using System.Text;
using System.Linq;

namespace AlgoTemplates
{
    public static class FFT
    {
		
		private static int GetClosestPower(int val)
		{
			return 1 << (int) Math.Ceiling(Math.Log(val, 2));
		}

		public static List<Complex> RunFFT(List<int> arr)
		{
			return RunFFT(arr.Select(el => (Complex)el).ToList(), 1);
		}

		public static List<Complex> RunFFT(List<Complex> arr)
		{
			return RunFFT(arr, 1);
		}

		public static List<Complex> RunIFFT(List<Complex> arr)
		{
			return RunFFT(arr, -1);
		}

		private static List<Complex> RunFFT(List<Complex> arr, int reverse)
		{
			var N = GetClosestPower(arr.Count);
			List<Complex> lst = new List<Complex>((int)N);
			foreach (var item in arr)
			{
				lst.Add(item);
			}
			for (int i = arr.Count; i < N; i++)
			{
				lst.Add(0);
			}

			internalFFT(lst, reverse);
			return lst;
		}

		static void internalFFT(List<Complex> arr, int reverse)
		{
			if (arr.Count == 1)
				return;

			var ev = arr.Where((el, i) => i % 2 == 0).ToList();
			var odd = arr.Where((el, i) => i % 2 == 1).ToList();

			internalFFT(ev, reverse);
			internalFFT(odd, reverse);
			var factor = (reverse *(2 * Math.PI)) / arr.Count;
			var powers = Enumerable.Range(0, arr.Count).Select(el => Complex.FromPolarCoordinates(1, factor * el)).ToList();

			for (int i = 0; i < arr.Count / 2; i++)
			{
				arr[i] = ev[i] + powers[i] * odd[i];
				arr[i + arr.Count / 2] = ev[i] - powers[i] * odd[i];
			}
		}

	}
}

File: AlgoTemplates/AlgoTemplatesCpp/FFT.cpp
#include "FFT.h"
#include <cmath>

using namespace std;

namespace AlgoTemplatesCpp
{
const double PI = 4 * atan(1);
int GetClosestPower(int val)
{
    return 1 << (int)ceil(log2(val));
}

void internalFFT(vector<complex<double>> &arr, int reverse)
{
    if (arr.size() == 1)
    {
        return;
    }

    vector<complex<double>> ev(arr.size() / 2), odd(arr.size() / 2), powers(arr.size());

    double factor = (reverse * (2 * PI)) / arr.size();
    for (int i = 0; i < arr.size() / 2; i++)
    {
        ev[i] = arr[2 * i];
        odd[i] = arr[2 * i + 1];
        powers[2 * i] = polar((double)1, factor * (2 * i));
        powers[2 * i + 1] = polar((double)1, factor * (2 * i + 1));
    }

    internalFFT(ev, reverse);
    internalFFT(odd, reverse);

    for (int i = 0; i < arr.size() / 2; i++)
    {
        arr[i] = ev[i] + powers[i] * odd[i];
        arr[i + arr.size() / 2] = ev[i] - powers[i] * odd[i];
    }
}

vector<complex<double>> RunFFT(const vector<int> &arr)
{
    int N = GetClosestPower(arr.size());
    vector<complex<double>> res(N);

    for (int i = 0; i < arr.size(); i++)
    {
        res[i] = arr[i];
    }
    internalFFT(res, 1);
    return res;
}

vector<complex<double>> RunFFT(const vector<complex<double>> &arr)
{
    int N = GetClosestPower(arr.size());
    vector<complex<double>> res(N);

    for (int i = 0; i < arr.size(); i++)
    {
        res[i] = arr[i];
    }
    internalFFT(res, 1);
    return res;
}

vector<complex<double>> RunIFFT(const vector<complex<double>> &arr)
{
    int N = GetClosestPower(arr.size());
    vector<complex<double>> res(N);

    for (int i = 0; i < arr.size(); i++)
    {
        res[i] = arr[i];
    }
    internalFFT(res, -1);
    return res;
}


    vector<complex<double>> Multiply(const vector<complex<double>>& a, const vector<complex<double>>& b)
    {     
        if(a.size() != b.size())  
            throw new exception();
        vector<complex<double>> res(a.size());
        for(int i =0; i< a.size(); i++)
        {
            res[i] = a[i] * b[i];
        }
        return res;

    }
} // namespace AlgoTemplatesCpp